nixosTests.nginx*: port to python

globin opened the pull request:
###### Motivation for this change
#72828 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

globin committed:
nixosTests.nginx*: port to python

flokli commented on nixos/tests/nginx.nix:
```diff
@@ -69,43 +69,46 @@ import ./make-test.nix ({ pkgs, ... }: {
+    webserver.wait_for_open_port("80")
+
+    with subtest("check ETag if serving Nix store paths"):
+        oldEtag = checkEtag
```
`checkEtag` is a function. I think this should be called
```py
old_etag = check_etag()
```

flokli commented on nixos/tests/nginx.nix:
```diff
@@ -69,43 +69,46 @@ import ./make-test.nix ({ pkgs, ... }: {
+        webserver.succeed(
+            "${etagSystem}/bin/switch-to-configuration test >&2"
+        )
+        webserver.sleep(1)  # race condition
```
What race condition happens here? We probably need to wait longer, just to be sure…

globin replied:
I've taken this out for now, as I can't reproduce the failure even on a build machine running over load.

flokli commented on nixos/tests/nginx.nix:
```diff
@@ -69,43 +69,46 @@ import ./make-test.nix ({ pkgs, ... }: {
+            "${etagSystem}/bin/switch-to-configuration test >&2"
+        )
+        webserver.sleep(1)  # race condition
+        newEtag = checkEtag
```
```suggestion
        new_etag = check_etag()
```

flokli commented on nixos/tests/nginx.nix:
```diff
@@ -69,43 +69,46 @@ import ./make-test.nix ({ pkgs, ... }: {
+        )
+        webserver.sleep(1)  # race condition
+        newEtag = checkEtag
+        assert oldEtag != newEtag
```
```suggestion
        assert old_etag != new_etag
```

flokli commented on nixos/tests/nginx.nix:
```diff
@@ -69,43 +69,46 @@ import ./make-test.nix ({ pkgs, ... }: {
+        webserver.succeed(
+            "${justReloadSystem}/bin/switch-to-configuration test >&2"
+        )
+        webserver.wait_for_open_port("8080")
```
```suggestion
        webserver.wait_for_open_port(8080)
```

globin committed:
nixosTests.nginx*: review fixes

Co-Authored-By: Florian Klink <flokli@flokli.de>

tfc commented on nixos/tests/nginx-sso.nix:
```diff
@@ -27,18 +27,20 @@ import ./make-test.nix ({ pkgs, ... }: {
+    machine.wait_for_open_port(8080)
 
     # No valid user -> 401.
-    $machine->fail("curl -sSf http://localhost:8080/auth");
```
the comments could be the titles of `subtest`s so this information is even visible in the log

Ma27 replied:
Done :)

tfc reviewed (changes requested):
Please make the comments with the http return codes subtest titles, that would be awesome.

flokli commented:
ping @globin

Ma27 committed:
nixos/tests/nginx-sso: add subtests for each assertion

Ma27 commented:
@GrahamcOfBorg test nginx nginx-sso

Ma27 committed:
nixos/nginx: fix test

When using format-strings, curly brackets need to be escaped using `{{`
to avoid errors from python.

And apparently, Perl's `==` is used to compare substrings[1] which is why
the translation to `assert http_code == "304"` failed as the string
contains several headers from curl.

[1] Just check `perl <(echo 'die "alarm" if "foo\n304" == 304')`

Ma27 commented:
Rebased onto latest master, fixed the `nginx` test.

@GrahamcOfBorg test nginx nginx-sso
